be/src/runtime/sender_queue.cpp

```diff
@@ -0,0 +1,957 @@
+                    break;
+                }
+            }
+        }
+    }
+    for (const auto& kv : _chunk_meta.tuple_id_to_index) {
+        _chunk_meta.types[kv.second] = TypeDescriptor(PrimitiveType::TYPE_BOOLEAN);
+        ++column_index;
+    }
+
+    if (UNLIKELY(column_index != _chunk_meta.is_nulls.size())) {
+        return Status::InternalError("build chunk meta error");
+    }
+    return Status::OK();
+}
+
+Status DataStreamRecvr::SenderQueue::_deserialize_chunk(const ChunkPB& pchunk, vectorized::Chunk* chunk,
+                                                        faststring* uncompressed_buffer) {
+    if (pchunk.compress_type() == CompressionTypePB::NO_COMPRESSION) {
+        SCOPED_TIMER(_recvr->_deserialize_chunk_timer);
+        TRY_CATCH_BAD_ALLOC({
+            serde::ProtobufChunkDeserializer des(_chunk_meta);
+            ASSIGN_OR_RETURN(*chunk, des.deserialize(pchunk.data()));
+        });
+    } else {
+        size_t uncompressed_size = 0;
+        {
+            SCOPED_TIMER(_recvr->_decompress_chunk_timer);
+            const BlockCompressionCodec* codec = nullptr;
+            RETURN_IF_ERROR(get_block_compression_codec(pchunk.compress_type(), &codec));
+            uncompressed_size = pchunk.uncompressed_size();
+            TRY_CATCH_BAD_ALLOC(uncompressed_buffer->resize(uncompressed_size));
+            Slice output{uncompressed_buffer->data(), uncompressed_size};
+            RETURN_IF_ERROR(codec->decompress(pchunk.data(), &output));
+        }
+        {
+            SCOPED_TIMER(_recvr->_deserialize_chunk_timer);
+            TRY_CATCH_BAD_ALLOC({
+                std::string_view buff(reinterpret_cast<const char*>(uncompressed_buffer->data()), uncompressed_size);
+                serde::ProtobufChunkDeserializer des(_chunk_meta);
+                StatusOr<vectorized::Chunk> res = des.deserialize(buff);
+                if (!res.ok()) return res.status();
+                *chunk = std::move(res).value();
+            });
+        }
+    }
+    return Status::OK();
+}
+
+DataStreamRecvr::NonPipelineSenderQueue::NonPipelineSenderQueue(DataStreamRecvr* parent_recvr, int32_t num_senders)
+        : SenderQueue(parent_recvr), _num_remaining_senders(num_senders) {}
+
+Status DataStreamRecvr::NonPipelineSenderQueue::get_chunk(vectorized::Chunk** chunk, const int32_t driver_sequence) {
+    std::unique_lock<Mutex> l(_lock);
+    // wait until something shows up or we know we're done
+    while (!_is_cancelled && _chunk_queue.empty() && _num_remaining_senders > 0) {
+        VLOG_ROW << "wait arrival fragment_instance_id=" << _recvr->fragment_instance_id()
+                 << " node=" << _recvr->dest_node_id();
+        _data_arrival_cv.wait(l);
+    }
+
+    if (_is_cancelled) {
+        return Status::Cancelled("Cancelled SenderQueue::get_chunk");
+    }
+
+    if (_chunk_queue.empty()) {
+        return Status::OK();
+    }
+
+    *chunk = _chunk_queue.front().chunk_ptr.release();
```

Comment:
Is it safe to release at here?

Reply:
> Is it safe to release at here?

These are all existing codes in original SenderQueue, and they should be fine.